### ISSUE TITLE
introduce PassThrough-only producer messages for conditional processing purposes

### ISSUE THREAD

**Issue by svezfaz** [akka/reactive-kafka#415]
An approach to solve the [conditional message processing](https://doc.akka.io/docs/akka-stream-kafka/current/atleastonce.html#conditional-message-processing) issue could be to introduce the concept of "PassThrough-only" message. Basically the `ProducerMessage` could be available in 2 flavours, one that carry a `ProducerRecord`, and one that doesn't.

Something like

```
sealed trait Message

final case class RecordMessage[K, V, +PassThrough](
    record: ProducerRecord[K, V],
    passThrough: PassThrough
) extends Message

final case class PassThroughMessage[+PassThrough](
    passThrough: PassThrough
) extends Message
```

`ProducerStage` could actually call `producer.send` in case of `RecordMessage`, and only pass them through in case of, well, `PassThroughMessage`.

This would allow conditional processing logic like the following

```
source
  .filter(predicate)
  .map(toProducerMessage)
  .via(Producer.flow)
  .to(commitSink)
```

to be changed to 

```
source
  .map {
    case (k, v, o) if predicate(x) => RecordMessage(k, v, o)
    case (_, _, o)                 => PassThroughMessage(o)
  }
  .via(Producer.flow)
  .to(commitSink)
```

where the latter doesn't skip committing filtered out elements.

**Comment by DWvanGeest** [comment]
Yes! I would like this as well.

My use case is that the source is actually another Kafka topic. Some messages can't be processed and re-published, and so are filtered out of the stream so as not to enter my `Producer.flow`. Since I commit _after_ the `Producer.flow`, those filtered messages are never committed. Instead, nothing is committed until a processable/produceable message comes through the stream.

**Comment by ennru** [comment]
Yes, I like this idea.
@svezfaz Would you like to propose a PR?

**Comment by schrepfler** [comment]
Might be good to have also message headers as well.

**Comment by svezfaz** [comment]
@ennru I'd be happy to work on a PR. How acceptable is it to break the scaladsl/javadsl flow contracts? Not sure this could be pulled off otherwise.

**Comment by ennru** [comment]
Not breaking APIs is better, of course. Give it a stab and let's see how it can be done.
For the new transactional support it does not matter, it hasn't been released, yet.

**Comment by ennru** [comment]
I started to work on this idea. We'll release a new version soon and I try to get this into the new transaction supporting flows. This is a chance to get it in without breaking any APIs - as it is new.

**Comment by svezfaz** [comment]
@ennru I think it's a good idea to try it out on the transactional flows. I gave it a stab by creating yet another flow (i.e. `flowConditional`) but wasn't really convinced by it..

**Comment by ennru** [comment]
Fixed by #495
